docs/cryptodoc/src/05_09_frodokem.rst

```diff
@@ -0,0 +1,298 @@
+
+   **Output:**
+
+   -  ``SK``, ``PK``: private and public key
+
+   **Steps:**
+
+   1. Generate new values ``s``, ``seed_se``, and ``z`` using ``rng``
+   2. ``seed_a = SHAKE(z, len_a)``
+   3. ``r = SHAKE(0x5F || seed_se, 32*n*n_bar)``
+   4. ``s_trans = sample_matrix(r[:n*n_bar - 1])``
+   5. ``e = sample_matrix(r[n*n_bar:])``
+   6. ``b = a*s + e``
+   7. | ``PK = {seed_a, pack(b)}``
+      | ``pkh = SHAKE(PK, len_sec)``
+      | ``SK = {s, seed_a, pack(b), s_trans, pkh}``
+
+   **Notes:**
+
+   - Computation of ``b = a*s + e`` is done by a specialised function that performs on-demand
+     expansion of ``seed_a`` into the desired row of the matrix :math:`A` of [FrodoKEM-ISO]_
+     and assumes getting the transpose ``s_trans`` of ``s`` as input to avoid transposition.
+   - The operation ``pack(b)`` is performed when accessing the serialized or raw key bits of
+     a key.
+   - The computation of ``pkh`` is performed in the constructor of ``FrodoKEM_PublicKeyInternal``,
+     an internal class used to represent ``PK``.
+   - The creation of a ``FrodoKEM_PublicKey`` is conducted using the
+     ``public_key`` method of the private key.
+
+..  _pubkey/frodokem/encapsulation:
+
+Key Encapsulation
+-----------------
+
+The FrodoKEM encapsulation procedure of Botan follows Section 8.2 of [FrodoKEM-ISO]_ and
+works as follows:
+
+.. admonition:: FrodoKEM Encapsulation
+
+   **Input:**
+
+   - ``PK = {seed_a, packed_b}``: public key
+   - ``rng``: random number generator
+
+   **Output:**
+
+   - ``encapsulated_key``: ciphertext of shared key
+   - ``shared_key``: plaintext shared key
+
+   **Steps:**
+
+   1. Generate new values ``u`` and ``salt`` using ``rng``
+   2.  ``seed_se || k = SHAKE(pkh || u || salt, len_se + len_sec)``
+   3. ``r = SHAKE(0x96 || seed_se, 16*(2*n_bar*n + n_bar*n_bar))``
+   4. ``s_p = sample_matrix(r[:n*n_bar - 1])``
+   5. ``e_p = sample_matrix(r[n*n_bar:2*n*n_bar - 1])``
+   6. ``b_p = s_p*a + e_p``
+   7. ``c_1 = pack(b_p)``
+   8. ``e_pp = sample_matrix(r[2*n*n_bar:])``
+   9.  ``v = s_p*b + e_pp``
+   10. ``c = v + encode(u)``
+   11. ``c_2 = pack(c)``
+   12. ``encapsulated_key = c_1 || c_2 || salt``
+   13. ``shared_key = SHAKE(encapsulated_key || k, len_sec)``
+
+   **Notes:**
+
+   - The computation of ``pkh`` is already performed in the constructor of ``FrodoKEM_PublicKeyInternal``,
+     an internal class used to represent ``PK``.
+   - ``b = unpack(packed_b)`` is already performed in the constructor of ``FrodoKEM_PublicKey``.
+   - The computation of ``s_p*a + e_p`` is done by a specialised function that performs on-demand
+     expansion of ``seed_a`` into the desired row of the matrix :math:`A` of [FrodoKEM-ISO]_.
+   - The computation of ``s_p*b + e_pp`` is done by a specialised function realising this sequence
+     of operations.
+
+..  _pubkey/frodokem/decapsulation:
+
+Key Decapsulation
+-----------------
+
+The FrodoKEM encapsulation procedure of Botan follows Section 8.3 of [FrodoKEM-ISO]_ and
```

Comment:
Can we give a link to the location in the code?

Reply:
We could, in fact we did so for Kyber and Dilithium. However, the text then became very cumbersome to read so based on prior feedback we introduced overview tables to see which logic is implemented in what code location. Probably related to https://github.com/sehlen-bsi/botan-docs/pull/117#discussion_r1328680313

Reply:
I think making FrodoKEM encapsulation in this line clickable to the location in the code where the algorithm is implemented would be helpful to certain readers while not inconveniencing others.

Reply:
I'm not against deep-linking to source locations. In fact, I do agree that this certainly is helpful to readers.

However, links to specific code locations always come with a certain maintenance burden. Whenever we refactor code (eg. https://github.com/randombit/botan/pull/3887), we need to remember and update these links in the documentation; especially when they point to a certain code location.

Given that the crypto documentation is meant to be a "living document" that gets updated as Botan evolves, links may become broken or stale. The central tables can be kept up-to-date fairly easily. Scattered links in the prose are more of a maintenance burden.

Perhaps we could extend our existing `:srcref:` rST-Extension to:
a) allow links to specific lines in the code (currently it supports linking to files only)
b) add an "expected text token", that can be checked automatically in CI

For instance (inventing some rST syntax here):

```
The FrodoKEM encapsulation (as implemented in :srcref:`src/lib/pubkey/frodokem/frodokem_common/frodokem.cpp#L89 <raw_key_encrypt>`) ...
```

In [the "linked" source location](https://github.com/randombit/botan/blob/7f15ce7b39d6bd34dda9af4d2cb08a6e18c64c01/src/lib/pubkey/frodokem/frodokem_common/frodokem.cpp#L89) we actually implement `raw_kem_encrypt`. A CI script could fetch the file and check that it (1) exists, (2) has a line "89" and (3) that line contains the string "raw_kem_encrypt". That should be good enough to detect stale deep links automatically and remove their maintenance burden.

Reply:
I completely understand that deep-linking comes with a lot of extra maintenance work, so I agree that manually inputting these links will not be sustainable in the long run. However, I think your suggested extension of the srcref-functionality to facilitate such "clever" linking would really improve the document if it is feasible to implement.

Reply:
`:srcref:` Functionality is implemented [here](https://github.com/sehlen-bsi/botan-docs/pull/188). Once merged, we can rebase this PR and use it here.

Reply:
I linked to the relevant locations in the upstream source file. Note that I added a specific (partial) path instead of just making the existing text clickable. reStructuredText has a limitation that makes is overly complicated to next backticked markups (i.e. \`\`FrodoKEM_PrivateKey`` inside a :srcref:\`text \<path>` block) 😞

Reply:
Since I accidentally created a typo in the link that was caught by CI: Here's how a misaligned deep-link would show up in CI:

![image](https://github.com/sehlen-bsi/botan-docs/assets/1562139/aa34383f-df74-45bc-9117-aebad1627ea4)

Reply:
FWIW: The one and only deep-link I added in #188 [got outdated already, CI detected it and I had to fix it](https://github.com/sehlen-bsi/botan-docs/pull/189#issuecomment-1952479493). ;)